test/runtests.jl

```diff
@@ -1,6 +1,63 @@
 using RealDot
+using LinearAlgebra
 using Test
 
+# struct need to be defined outside of tests for julia 1.0 compat
+# custom complex number (tests fallback definition)
+struct CustomComplex{T} <: Number
+    re::T
+    im::T
+end
+
+Base.real(x::CustomComplex) = x.re
+Base.imag(x::CustomComplex) = x.im
+
+Base.conj(x::CustomComplex) = CustomComplex(x.re, -x.im)
+
+function Base.:*(x::CustomComplex, y::Union{Real,Complex})
+    return CustomComplex(reim(Complex(reim(x)...) * y)...)
+end
+Base.:*(x::Union{Real,Complex}, y::CustomComplex) = y * x
+function Base.:*(x::CustomComplex, y::CustomComplex)
+    return CustomComplex(reim(Complex(reim(x)...) * Complex(reim(y)...))...)
+end
+
+# custom quaternion to test definition for hypercomplex numbers
+# adapted from Quaternions.jl
+struct Quaternion{T<:Real} <: Number
+    s::T
+    v1::T
+    v2::T
+    v3::T
+end
+
+Base.real(q::Quaternion) = q.s
+Base.conj(q::Quaternion) = Quaternion(q.s, -q.v1, -q.v2, -q.v3)
+
+function Base.:*(q::Quaternion, w::Quaternion)
+    return Quaternion(
+        q.s * w.s - q.v1 * w.v1 - q.v2 * w.v2 - q.v3 * w.v3,
+        q.s * w.v1 + q.v1 * w.s + q.v2 * w.v3 - q.v3 * w.v2,
+        q.s * w.v2 - q.v1 * w.v3 + q.v2 * w.s + q.v3 * w.v1,
+        q.s * w.v3 + q.v1 * w.v2 - q.v2 * w.v1 + q.v3 * w.s,
+    )
+end
+
+function Base.:*(q::Quaternion, w::Union{Real,Complex,CustomComplex})
+    a, b = reim(w)
+    return q * Quaternion(a, b, zero(a), zero(a))
+end
+Base.:*(q::Union{Real,Complex,CustomComplex}, w::Quaternion) = w * q
```

Comment:
Haha yeah I guess it's clear that I've never worked with quaternions :smile: In fact for the tests it does not matter and the result could be anything since `Quaternion` will use the fallback definition `realdot(x, y) = real(dot(x, y))` and hence tests will always pass, regardless of whether `dot` is implemented correctly or not, as long as some value can be computed. Of course, I'll fix it so that it seems that I know what I am doing here even though I just copy stuff from Quaternions :smile:

Reply:
Yeah, quaternions are super useful as a sanity check for chain rules of numers/numerical arrays, since they behave like matrices. I have probably thought way too much about them.